src/snowflake/connector/util_text.py

```diff
@@ -33,6 +43,109 @@ def split_statements(
     buf: StringIO,
     remove_comments: bool = False,
     delimiter: Optional[SQLDelimiter] = None,
+):
+    """Entry function to split statements."""
+    if delimiter is None or str(delimiter) == ";":
+        return split_statements_new(buf, remove_comments)
+    else:
+        return split_statements_old(buf, remove_comments, delimiter)
+
+
+def split_statements_new(
+    buf: StringIO,
+    remove_comments: bool = False,
+):
+    """split_statements_new is not taking delimiter.
+    This new method tries to use a mini ANTLR parser to recognize SQL statement blocks, especially
+    'create procedure...begin .. end'
+    In this method, users could also use manual way to guide the splitting
+    by using special comments. A pure comment line '--SF:<snowflake splitting hints>'
+    All these commands should not be nested
+      (1) '--SF:auto_split off': disable auto split. This should be first cmd in a script file
+      (2) to disable split the sql texts between these two hints
+      '--SF:<<'
+         ...
+      '--SF:>>'
+      (3) to split the blocks before and after this comment '--SF:<<>>'
+    """
+    sqltext = buf.read()
+    sfcmds = []
+    for m in SF_COMMENT_CMD_PATTERN.finditer(sqltext.lower()):
+        sfcmds.append([m.start(), m.end()])
+
+    # stmts to store the split result to return
+    stmts = []
+    is_put_get = []
+    is_auto_split = True
+
+    sqltext_lower = sqltext.lower()
+    for cmd in sfcmds:
+        if is_auto_split and RE_SPLIT_OFF.match(sqltext_lower, cmd[0], cmd[1]):
+            _logger.info("auto_split off")
+            is_auto_split = False
+    starting_pos_of_block = -1  # not in block yet
+    last_pos = 0  # last pos of ending blocks
+
+    # pieces contains tuples [(0|1, str)].
```

Comment:
Instead of this comment, type-hint it.
Maybe even define a `NamedTuple` for the individual elements could be done. I find this documentation to be much more impactful than these random comments in the middle of a function